test/functional/p2p_addr_relay.py

```diff
@@ -82,6 +130,49 @@ def run_test(self):
         ipv4_branching_factor = 2
         assert_equal(total_ipv4_received, num_ipv4_addrs * ipv4_branching_factor)
 
+        self.nodes[0].disconnect_p2ps()
+
+    def inbound_blackhole_tests(self):
+        self.log.info('Check that we only relay addresses to inbound peers who have previously sent us addr related messages')
+
+        addr_source = self.nodes[0].add_p2p_connection(P2PInterface())
+        receiver_peer = self.nodes[0].add_p2p_connection(AddrReceiver())
+        blackhole_peer = self.nodes[0].add_p2p_connection(AddrBlackhole())
+
+        msg = self.setup_addr_msg(2)
+        self.send_addr_msg(addr_source, msg, [receiver_peer, blackhole_peer])
+
+        assert_equal(receiver_peer.num_ipv4_received, 2)
+        assert_equal(blackhole_peer.num_ipv4_received, 0)
+
+        self.nodes[0].disconnect_p2ps()
```

Comment:
> Since peers are chosen randomly..

yes, peers are chosen randomly, but this test is set up to not be random- ipv4 addresses will be relayed to 2 peers that are not the source of the addr message. so both `receiver_peer` and `blackhole_peer` would receive the addr if they were eligible. definitely wouldn't want a flaky test there :) 

> I think this test could also be improved.. 

thats a great idea! I've added locally, will include in next push. 

> assert bytes received for addr-related messages = 0

hm, `blackhole_peer` is a p2p connection, so we can't call `getpeerinfo()` on it to get the bytes received directly. that said, we could check `bytessent_per_msg` on the node to see that no bytes have been sent for an `addr` message. however, I its a bit brittle & the p2p connection is capturing directly receiving the addr message, so I don't think that's adding much value to the test. let me know if you think it'd add test coverage :)

Reply:
> yes, peers are chosen randomly, but this test is set up to not be random- ipv4 addresses will be relayed to 2 peers that are not the source of the addr message. so both receiver_peer and blackhole_peer would receive the addr if they were eligible. definitely wouldn't want a flaky test there :)

Oh right it's always going to be the same one. That makes sense then, that the node has 2 options but doesn't send it to a peer that hasn't established addr relay. But wait I'm confused why it would be flaky, I thought `blackhole_peer` would never get the addr because it hasn't sent any addr-related messages?

> hm, blackhole_peer is a p2p connection, so we can't call getpeerinfo() on it to get the bytes received directly.

Er I mean, call `getpeerinfo()` on the node to see bytes received, kinda like this:

```py
def sum_addr_messages(msgs):
    return msgs['addr'] + msgs['addrv2'] + msgs['getaddr'] + msgs['sendaddrv2']
...

peerinfo = self.nodes[0].getpeerinfo()
# receiver_peer sent us its own addr and a getaddr after the VERSION message, so we relay
assert sum_addr_messages(peerinfo[1]['bytesrecv_per_msg']) > 0
assert_equal(receiver_peer.num_ipv4_received, 2)

# blackhole_peer should not have sent us any addr-related messages, so we don't relay
assert sum_addr_messages(peerinfo[2]['bytesrecv_per_msg']) == 0
assert_equal(blackhole_peer.num_ipv4_received, 0)
```
I just think it's a more precise test, asserting that the addrs aren't relayed because no addr messages have been received. If addr dialogue changes, this test catches it/debugs itself. Not sure if that counts as coverage though

Reply:
> But wait I'm confused why it would be flaky, I thought blackhole_peer would never get the addr because it hasn't sent any addr-related messages?

yes yes, it would be flaky if it wasn't deterministic / `blackhole_peer` was being skipped by chance. I think we're on the same page, just confusion of my language now. 

> Er I mean, call getpeerinfo() on the node to see bytes received, kinda like this

oh I get it now, you're suggesting additional coverage to test the setup, not another way to test the `addr` relayed. that's a cool idea! sure, I'll add it :) 

smol note: 
> # receiver_peer sent us its own addr and a getaddr after the VERSION message, so we relay

since `receiver_peer` is essentially a `P2PInterface` object, it actually won't self announce upon receiving `version` message. I added  that it will send out a `getaddr` so tests wouldn't unexpectedly break  in this patch. but I think that highlights why this suggested coverage is useful, at the very least it communicates the relevant setup that's otherwise "invisible" from the POV of this test file.

Reply:
based on this convo, I've updated the tests to assert that `blackhole_peer` starts receiving addr messages after sending one. I've also added `getpeerinfo` for the node to verify the test setup of which peers have sent addr related messages. 

thanks!